File: cmd/sieve/sieve.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

void sieve_of_eratosthenes(int n) {
    if (n < 2) {
        printf("Prime numbers up to %d:\n", n);
        printf("\nTotal: 0 primes\n");
        return;
    }

    // Create a boolean array "is_prime[0..n]" and initialize all entries as true
    bool *is_prime = (bool *)malloc((n + 1) * sizeof(bool));
    if (is_prime == NULL) {
        fprintf(stderr, "Error: memory allocation failed\n");
        exit(1);
    }

    memset(is_prime, true, (n + 1) * sizeof(bool));
    is_prime[0] = false;
    is_prime[1] = false;

    // Start with the smallest prime number, 2
    for (int p = 2; p * p <= n; p++) {
        // If is_prime[p] is not changed, then it is a prime
        if (is_prime[p]) {
            // Mark all multiples of p as not prime
            for (int i = p * p; i <= n; i += p) {
                is_prime[i] = false;
            }
        }
    }

    // Print all numbers that are still marked as prime
    printf("Prime numbers up to %d:\n", n);
    int count = 0;
    for (int i = 2; i <= n; i++) {
        if (is_prime[i]) {
            if (count > 0 && count % 10 == 0) {
                printf("\n");
            }
            printf("%d ", i);
            count++;
        }
    }
    printf("\n\nTotal: %d primes\n", count);

    free(is_prime);
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: sieve <n>\n");
        printf("Finds all prime numbers up to n using the Sieve of Eratosthenes\n");
        return 1;
    }

    char *endptr;
    long n = strtol(argv[1], &endptr, 10);

    if (*endptr != '\0' || endptr == argv[1]) {
        fprintf(stderr, "Error: invalid number: %s\n", argv[1]);
        return 1;
    }

    if (n < 0) {
        fprintf(stderr, "Error: n must be non-negative\n");
        return 1;
    }

    if (n > 2147483647) {
        fprintf(stderr, "Error: n is too large\n");
        return 1;
    }

    sieve_of_eratosthenes((int)n);

    return 0;
}

File: cmd/sieve/sieve.zig
const std = @import("std");

/// Generate all prime numbers up to n using the Sieve of Eratosthenes.
fn sieveOfEratosthenes(allocator: std.mem.Allocator, n: usize) !std.array_list.Managed(usize) {
    var primes = std.array_list.Managed(usize).init(allocator);
    errdefer primes.deinit();

    if (n < 2) {
        return primes;
    }

    // Create a boolean array "is_prime" and initialize all entries as true
    const is_prime = try allocator.alloc(bool, n + 1);
    defer allocator.free(is_prime);

    @memset(is_prime, true);
    is_prime[0] = false;
    is_prime[1] = false;

    // Start with the smallest prime number, 2
    var p: usize = 2;
    while (p * p <= n) : (p += 1) {
        // If is_prime[p] is not changed, then it is a prime
        if (is_prime[p]) {
            // Mark all multiples of p as not prime
            var i: usize = p * p;
            while (i <= n) : (i += p) {
                is_prime[i] = false;
            }
        }
    }

    // Collect all numbers that are still marked as prime
    var i: usize = 2;
    while (i <= n) : (i += 1) {
        if (is_prime[i]) {
            try primes.append(i);
        }
    }

    return primes;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    const stderr = std.fs.File.stderr();
    const stdout = std.fs.File.stdout();

    if (args.len < 2) {
        _ = try stderr.write("Usage: sieve-zig <n>\n");
        _ = try stderr.write("Finds all prime numbers up to n using the Sieve of Eratosthenes\n");
        std.process.exit(1);
    }

    const n = std.fmt.parseInt(usize, args[1], 10) catch {
        var buf: [256]u8 = undefined;
        const msg = try std.fmt.bufPrint(&buf, "Error: invalid number: {s}\n", .{args[1]});
        _ = try stderr.write(msg);
        std.process.exit(1);
    };

    const primes = try sieveOfEratosthenes(allocator, n);
    defer primes.deinit();

    // Print header
    var buf: [256]u8 = undefined;
    const header = try std.fmt.bufPrint(&buf, "Prime numbers up to {d}:\n", .{n});
    _ = try stdout.write(header);

    // Print primes with line breaks every 10 numbers
    for (primes.items, 0..) |prime, i| {
        if (i > 0 and i % 10 == 0) {
            _ = try stdout.write("\n");
        }
        const prime_str = try std.fmt.bufPrint(&buf, "{d} ", .{prime});
        _ = try stdout.write(prime_str);
    }

    if (primes.items.len > 0) {
        _ = try stdout.write("\n");
    }

    const footer = try std.fmt.bufPrint(&buf, "\nTotal: {d} primes\n", .{primes.items.len});
    _ = try stdout.write(footer);
}
